gh-97696: asyncio eager tasks factory

itamaro opened the pull request:
fixes gh-97696

pyperformance results:
* async tree benchmarks up to ~50% faster when using the eager tasks factory (depending on how many leaf nodes actually suspend)
* geometric mean for all benchmark 1% faster

pyperformance run performed with local patch that installs the eager tasks factory by default (modifying the benchmark itself to use it requires https://github.com/psf/pyperf/pull/157)

full comparison in [this gist](https://gist.github.com/itamaro/106f100fc7753ac1960c696893fb2b47)

itamaro committed:
Eager task factory implementation

itamaro committed:
Eager task factory tests

itamaro committed:
Add NEWS and docs for eager task factory

carljm commented on Doc/library/asyncio-task.rst:
```diff
@@ -518,6 +518,30 @@ Running Tasks Concurrently
+.. function:: create_eager_task_factory(custom_task_constructor)
+
+    Create an eager task factory, similar to :func:`eager_task_factory`,
+    using the provided *eager_task_factory* when creating a new task instead
```
```suggestion
    using the provided *custom_task_constructor* when creating a new task instead
```

gvanrossum commented on Doc/library/asyncio-task.rst:
```diff
@@ -518,6 +518,30 @@ Running Tasks Concurrently
+    coroutines that are able to complete synchronously (without suspending)
+    are returned immediately as a completed :class:`Future`.
+
+    A regular :class:`Task` is returned otherwise, at the first suspension of *coro*.
```
This probably needs more/clearer explanation. I'd explain that it always tries to execute the coroutine immediately, until it either blocks, returns, or raises, and that when it blocks it returns a Task, when it returns/raises it returns a Future.

Returning a Future is still something I am skeptical about, since the task factory is global (for everything using the same event loop); I'll have to re-read the Issue. I'll come back here for the rest of the code review after I've had time for that.

itamaro committed:
elaborate explanation in docs and add a whatsnew entry

itamaro committed:
fix docs

itamaro committed:
Overhaul eager task factory design

Always create a task object to resolve the design flaw with structured concurrency.
Push the eager step execution into the task implementation, so the task itself knows to eagerly execute the first step if created via the eager task factory.

Co-authored-by: Jacob Bower <jbower@meta.com>

itamaro committed:
Fix task-counting tests for new impl

itamaro committed:
Fix test_task_exc_handler_correct_context

Co-authored-by: Jacob Bower <jbower@meta.com>

itamaro committed:
Merge branch 'main' into eager-tasks-factory

itamaro committed:
Merge branch 'main' into eager-tasks-factory

itamaro committed:
add jbower credit

itamaro committed:
cleanup recursive_taskgroups test case

itamaro committed:
Merge branch 'main' into eager-tasks-factory

itamaro committed:
Update asyncio documentation with latest state of the PR

itamaro committed:
don't add coro to the task repr if coro is None

itamaro committed:
also update the NEWS entry

itamaro committed:
add error check when using _PyDict_GetItem_KnownHash in swap_current_task

itamaro committed:
focus the eager task factory test suite on testing eager execution semantics

itamaro committed:
ensure task_eager_start is not called with a NULL task

jbower-fb commented on Misc/NEWS.d/next/Library/2023-03-15-12-18-07.gh-issue-97696.DtnpIC.rst:
```diff
@@ -0,0 +1,6 @@
+Implemented an eager task factory in asyncio.
+When used as a task factory on an event loop, it performs eager execution of
+coroutines. coroutines that are able to complete synchronously (e.g. return or
```
```suggestion
coroutines. Coroutines that are able to complete synchronously (e.g. return or
```

jbower-fb commented on Lib/test/test_asyncio/test_eager_task_factory.py:
```diff
@@ -0,0 +1,241 @@
+            return await fut
+
+        result = self.loop.run_until_complete(run())
+        self.assertEqual(result,'my message')
```
I'm not sure what this is testing beyond the `test_eager_completion()` below? I wouldn't expect `Future`s to behave differently in the case of eager execution.

itamaro replied:
I wouldn't expect Future to behave differently - so let's test that that's indeed the case.

I'm not 100% sure this is testing exactly what @gvanrossum asked me about "awaiting a future during the eager step", so hopefully he can chime in here.

jbower-fb commented on Lib/test/test_asyncio/test_eager_task_factory.py:
```diff
@@ -0,0 +1,241 @@
+            return 'hello'
+
+        async def run():
+            await asyncio.sleep(0.1)
```
I'm not sure this adds anything. What's your intent with this sleep (and the others also in tests' `run()` functions)?

itamaro replied:
in some cases I needed this to make sure the event loop started before doing anything that expects to take an eager step, because the eager step is taken *only* if the event loop is already running.

I refactored the test code to make this clearer, and removed the sleep where it wasn't actually needed.

jbower-fb commented on Lib/test/test_asyncio/test_eager_task_factory.py:
```diff
@@ -0,0 +1,241 @@
+                raise
+
+        with self.assertRaises(asyncio.CancelledError) as cm:
+            self.loop.run_until_complete(run())
```
I don't think we need to re-raise + catch the `CancelledError`?

```suggestion
            with self.assertRaises(asyncio.CancelledError):
                result = await t
            self.assertTrue(t.cancelled())

        self.loop.run_until_complete(run())
```

jbower-fb commented on Lib/test/test_asyncio/test_eager_task_factory.py:
```diff
@@ -0,0 +1,241 @@
+    )
+
+
+class BaseTaskCountingTests:
```
Can this derive from `test_utils.TestCase`, allowing us to skip the multiple inheritance for `NonEagerTests` etc.?

itamaro replied:
not really.

it's the same pattern as in other similar cases (like tasks tests), and the reasoning is that the base class has `test_*` methods but if it inherits from a `TestCase` then these methods will be executed as tests on the base class which fails (not surprisingly).

jbower-fb commented on Modules/_asynciomodule.c:
```diff
@@ -3053,6 +3147,65 @@ task_step(asyncio_state *state, TaskObj *task, PyObject *exc)
+static int
+task_eager_start(asyncio_state *state, TaskObj *task)
+{
+    assert(task != NULL);
```
This is fine, just curious why you added it?

itamaro replied:
to make sure we never call this function with task==NULL - it's an implicit assumption in the function body, this makes the assumption explicit

jbower-fb commented on Doc/library/asyncio-task.rst:
```diff
@@ -527,6 +527,41 @@ Running Tasks Concurrently
+    returns, or raises.
+    If the coroutine returns or raises, a finished :class:`Task` is returned,
+    and the task is never scheduled to the event loop. If the coroutine blocks,
+    the (pending) :class:`Task` is scheduled and returned.
```
```suggestion
    When using this factory (via ``loop.set_task_factory(asyncio.eager_task_factory)``),
    coroutines begin execution synchronously on `Task` construction. `Task`s are only
    scheduled  on the event loop if they block. This can be a performance improvement as
    the overhead  of loop scheduling is avoided for coroutines that complete synchronously.
    A common  example where this is beneficial is coroutines which employ caching to avoid
    actual I/O where possible.
```

willingc commented on Doc/library/asyncio-task.rst:
```diff
@@ -527,6 +527,41 @@ Running Tasks Concurrently
+    coroutines that are able to complete synchronously (without blocking)
+    are returned immediately as a finished :class:`Task`.
+
+    This task factory attempts to execute the coroutine ``coro`` immediately
```
```suggestion
    The eager task factory attempts to execute the coroutine ``coro`` immediately before
    scheduling the task to the event loop. The coroutine will execute until it blocks, returns,
    or raises. If the coroutine blocks, the pending :class:`Task` is scheduled to the event loop
    and returned. If coroutine returns or raises, a finished :class:`Task` is returned, and the task
    is not scheduled to the event loop.
    
```

itamaro replied:
@willingc does the latest version of this paragraph look good to you?

willingc replied:
Yup. Looks good @itamaro.

willingc commented on Doc/library/asyncio-task.rst:
```diff
@@ -527,6 +527,41 @@ Running Tasks Concurrently
+
+    .. note::
+
+        The fact that the coroutine starts execution immediately is a semantic change,
```
Immediate execution of the coroutine is a semantic change. If the coroutine blocks, the task is never scheduled to the event loop. If the coroutine execution returns or raises, the task is scheduled to the event loop. This change may introduce behavior changes to existing applications. For example, the application's task execution order is likely to change.

itamaro committed:
Refactor eager task tests to clarify the "loop is running" constraint and make it explicit

itamaro committed:
Apply documentation suggestions and feedback

itamaro committed:
Merge remote-tracking branch 'upstream/main' into eager-tasks-factory

itamaro committed:
fix docs (rst is hard)

itamaro committed:
Merge branch 'main' into eager-tasks-factory

jbower-fb committed:
Extend eager task factory tests

Added new tests for: checking current task + all_tasks, and context
vars.

Also made EagerTaskFactoryLoopTests run for both C and Python Task
implementations.

itamaro committed:
a little cleanup of newly added tests

itamaro committed:
add assertion to current_task test, comparing the task before and after eager stop blocking

itamaro committed:
add a second step to contextvars test

itamaro committed:
Merge branch 'main' into eager-tasks-factory

willingc committed:
Merge branch 'main' into eager-tasks-factory

willingc reviewed (approved):
Thanks for the updates @itamaro. This looks good to me but would love someone else to also take a look.

itamaro committed:
Merge branch 'main' into eager-tasks-factory

gvanrossum reviewed (approved):
LGTM (assuming the refcount leak buildbots are happy). Go ahead without waiting for me with further changes!

PS. Next time please resist the urge to rebase -- it's harder to pick up a review after a rebase has happened. We squash commits when we merge them into main anyways.

carljm commented on Doc/library/asyncio-task.rst:
```diff
@@ -527,6 +527,42 @@ Running Tasks Concurrently
+        Immediate execution of the coroutine is a semantic change.
+        If the coroutine blocks, the task is never scheduled to the event loop.
+        If the coroutine execution returns or raises, the task is scheduled to
+        the event loop. This change may introduce behavior changes to existing
```
Am I missing something, or is this backwards? With eager execution, if the coroutine blocks, the task will be scheduled to the event loop; if the coroutine returns or raises, it will never be scheduled to the event loop.

itamaro replied:
good catch! yes this is backwards.

I was also thinking to add a note to the note about the "order of task execution" being arbitrary and not guaranteed regardless -- that's true, right?

carljm replied:
Unfortunately I'm not enough of an asyncio expert to know off-hand what guarantees are or aren't currently made about the order of task execution. I didn't find any such guarantees in a quick skim of the docs, but I might have just missed the relevant section.

carljm commented on Lib/asyncio/tasks.py:
```diff
@@ -43,22 +44,25 @@ def all_tasks(loop=None):
-            tasks = list(_all_tasks)
+            if scheduled_tasks is None:
+                scheduled_tasks = list(_scheduled_tasks)
+            eager_tasks = list(_eager_tasks)
```
Any particular reason we avoid re-casting `_scheduled_tasks`, but don't bother with that check for `_eager_tasks`?

itamaro replied:
I think all this iteration is needed only for the WeakSet (`_scheduled_tasks`)

the potential issue is that an eager task will be converted to a regular task in another thread, after the scheduled tasks WeakSet was casted to a list, but before the eager task was captured, so it would be missing from the return value.

I think this can be fixed safely by capturing the eager tasks as a list above (before starting the iterations) instead of in the loop. this way, if an eager task becomes a regular task, it will be included twice (and then deduplicated in the set comprehension).

carljm replied:
Yes, makes sense that the "other thread modifies" problem is only an issue for `WeakSet` and not for `set` (which should be thread-safe.) Given this, I think the comment above here shouldn't be modified as the PR currently modifies it to suggest that it's an issue for both.

Pulling the case of `_eager_tasks` outside the loop makes sense. Do we even have to cast it to list at all?

itamaro replied:
I think we need a copy for thread safety. whether the copy is a list or a set doesn't really matter (maybe list is more efficient?)

how does it look like now?

carljm replied:
Updated code looks good to me, modulo one tiny grammar nit

carljm commented on Misc/NEWS.d/next/Library/2023-03-15-12-18-07.gh-issue-97696.DtnpIC.rst:
```diff
@@ -0,0 +1,6 @@
+coroutines. Coroutines that are able to complete synchronously (e.g. return or
+raise without blocking) are returned immediately as a finished task, and the
+task is never scheduled to the event loop. If the coroutine blocks, the
+(pending) is scheduled and returned.
```
seems like a noun is missing here. "the (pending) _task_ is scheduled and returned."

carljm reviewed (commented):
LGTM barring a couple doc nits and one question.

itamaro committed:
missing word in NEWS entry

itamaro committed:
refactor all_tasks() handling of eager_tasks

itamaro committed:
fix docs (PR review)

itamaro committed:
move cv.set inside the main task (otherwise it refleaks)

itamaro committed:
Merge branch 'main' into eager-tasks-factory

carljm commented on Lib/asyncio/tasks.py:
```diff
@@ -43,22 +44,25 @@ def all_tasks(loop=None):
+    # a regular task in another thread, we don't risk missing it
+    eager_tasks = list(_eager_tasks)
+    # Looping over the weak set isn't safe as it can be updated from another thread,
+    # therefore we cast to lists prior to filtering. The list cast itself requires
```
```suggestion
    # therefore we cast to list prior to filtering. The list cast itself requires
```

itamaro committed:
fix grammar in all_tasks comments

itamaro committed:
Merge branch 'main' into eager-tasks-factory

kumaraditya303 commented on Modules/_asynciomodule.c:
```diff
@@ -3062,6 +3156,65 @@ task_step(asyncio_state *state, TaskObj *task, PyObject *exc)
+        }
+    } else {
+        // This seems to really help performance on pyperformance benchmarks
+        Py_CLEAR(task->task_coro);
```
Is this documented anywhere? I can't find it

itamaro replied:
@jbower-fb

jbower-fb replied:
See my comments on https://github.com/python/cpython/issues/97696.

jbower-fb replied:
Also https://github.com/python/cpython/pull/104304 as possible documentation.

kumaraditya303 commented on Lib/asyncio/tasks.py:
```diff
@@ -897,17 +930,41 @@ def callback():
-_all_tasks = weakref.WeakSet()
+def create_eager_task_factory(custom_task_constructor):
+
+    if "eager_start" not in inspect.signature(custom_task_constructor).parameters:
```
I am strong -1 on using any introspection, this is wrong as will reject any factory whose signature cannot be represented by inspect module. This needs to be fixed.

jbower-fb replied:
I added this, but have no particular love for this code. I actually thought we already removed it as it caused some issues when the factory is a function with variable kwargs.

kumaraditya303 commented on Modules/_asynciomodule.c:
```diff
@@ -8,6 +8,7 @@
 #include "pycore_runtime_init.h"  // _Py_ID()
 #include "pycore_moduleobject.h"  // _PyModule_GetState()
 #include "structmember.h"         // PyMemberDef
+#include "cpython/context.h"
```
Isn't this is already included by `Python.h`?

jbower-fb replied:
Whoops, sorry about that. There is some plugin in my editor which mysteriously likes to add `#includes` which aren't needed. I've not had success disabling it!

kumaraditya303 commented on Modules/_asynciomodule.c:
```diff
@@ -2015,6 +2033,54 @@ leave_task(asyncio_state *state, PyObject *loop, PyObject *task)
+{
+    PyObject *func = PyObject_GetAttr(loop, &_Py_ID(is_running));
+    if (func == NULL) {
+        PyErr_Format(PyExc_TypeError, "Loop missing is_running()");
```
It would be much better to take advantage of load method optimization here and avoid creating bound method here.

gvanrossum replied:
(I.e., use `PyObject_CallMethodNoArgs()`)
